sources/Launcher/DYNCriticalTimeLauncher.cpp

```diff
@@ -0,0 +1,130 @@
+//
+// Copyright (c) 2021, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source suite
+// of simulation tools for power systems.
+//
+
+/**
+ * @file  CriticalTimeLauncher.cpp
+ *
+ * @brief Critical Time launcher: implementation of the algorithm and interaction with dynawo core
+ *
+ */
+
+#ifdef WITH_OPENMP
+#include <omp.h>
+#endif
+
+#include <DYNExecUtils.h>
+#include <DYNSimulation.h>
+#include <DYNSubModel.h>
+#include <DYNModelMulti.h>
+#include "DYNCriticalTimeLauncher.h"
+#include "DYNMultipleJobs.h"
+#include "DYNAggrResXmlExporter.h"
+#include "MacrosMessage.h"
+
+using multipleJobs::MultipleJobs;
+
+namespace DYNAlgorithms {
+
+void
+CriticalTimeLauncher::launch() {
+  const std::string& jobsFile = criticalTimeCalculation_->getJobsFile();
+
+#ifdef WITH_OPENMP
+omp_set_num_threads(nbThreads_);
+#endif
+
+  inputs_.readInputs(workingDirectory_, jobsFile, 1);
+  boost::shared_ptr<job::JobEntry> job = inputs_.cloneJobEntry();
+  SimulationParameters params;
+  boost::shared_ptr<DYN::Simulation> simulation = createAndInitSimulation(workingDirectory_, job, params, results_, inputs_);
+  
+  if (simulation) {
+    boost::shared_ptr<DYN::ModelMulti> modelMulti = boost::dynamic_pointer_cast<DYN::ModelMulti>(simulation->model_);
+    std::string DDBDir = getEnvVar("DYNAWO_DDB_DIR");
+
+    const std::string& dydId = criticalTimeCalculation_->getDydId();
+    const std::string& endPar = criticalTimeCalculation_->getEndPar();
+
+    subModel_ = modelMulti->findSubModelByName(dydId);
+    subModel_->setParameterValue(endPar, DYN::PAR, tSup_, false);
+    subModel_->setSubModelParameters();
+    
+    simulate(simulation, results_);
+  }
+}
+
+void
+CriticalTimeLauncher::updateIndexes(double& tPrevious, double& curAccuracy, const double& multiplierRound) {
+  curAccuracy = std::abs(tSup_ - tPrevious);
+  double midDichotomy = std::round((std::abs(tSup_ - tPrevious) / 2) * multiplierRound) / multiplierRound;
+  double tmp = tSup_;
+
+  if(results_.getSuccess())
+    tSup_ += midDichotomy;
+  else 
+    tSup_ -= midDichotomy;
+
+  tPrevious = tmp;
+}
+
+void
+CriticalTimeLauncher::SearchCriticalTime() {
```

Comment:
This way of doing things breaks the architecture or the launchers and forces the main to have a dedicated call. Please use launch for the main algorithm and create a "SetParametersAndLaunchSimulation" method to launch a simulation. Moreover we read N times the jobs file whereas it is always the same, why not keeping the object in memory?